src/openzaak/utils/validators.py

```diff
@@ -115,3 +117,52 @@ def __call__(self, value: str):
             )
 
         return obj
+
+
+class ObjecttypeInformatieobjecttypeRelationValidator:
+    code = "missing-{}-informatieobjecttype-relation"
+    message = _("Het informatieobjecttype hoort niet bij het {} van de {}.")
+
+    def __init__(self, object_field: str = "zaak", objecttype_field: str = "zaaktype"):
+        self.object_field = object_field
+        self.objecttype_field = objecttype_field
+
+    def __call__(self, attrs):
+        code = self.code.format(self.objecttype_field)
+        message = self.message.format(self.objecttype_field, self.object_field)
+
+        informatieobject = attrs.get("informatieobject")
+        object = attrs.get(self.object_field)
+        if not informatieobject or not object:
+            return
+
+        objecttype = getattr(object, self.objecttype_field)
+
+        if not isinstance(informatieobject, EnkelvoudigInformatieObject):
+            io_type = informatieobject.latest_version.informatieobjecttype
+        else:
+            io_type = informatieobject.informatieobjecttype
+
+        # zaaktype/besluittype and informatieobjecttype should be both internal or external
+        if bool(objecttype.pk) != bool(io_type.pk):
+            raise serializers.ValidationError(message, code=code)
```

Comment:
Maybe I'm nitpicking, but wouldn't it be a bit better to have a different error message here that states that both types should be either internal or external, instead of just returning the general message `The informatieobjecttype is not related to the zaaktype of the zaak.`?

Reply:
Honestly I don't know what to write in the message in this case to make it clear from the client's point of view. Maybe something like this:
`The informatieobjecttype and the zaaktype should belong to the same component` ?

Reply:
Maybe something like `The informatieobjecttype and the zaaktype should be hosted in the same component`?

Reply:
Yeah, this version is better. The message is changed

Reply:
I've changed the error message into specifying that both types must belong to the same catalogue, I think that's the semantic meaning that is expressed behind this, and that requirement is specified in the catalogue API as well.